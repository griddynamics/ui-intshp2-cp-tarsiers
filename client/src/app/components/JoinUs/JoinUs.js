import React, { Component } from 'react';

import Snackbar from '../Snackbar/Snackbar';
import appConfig from '../../../config/appConfig';
import './JoinUs.scss';

class JoinUs extends Component {
  snackbarRef = React.createRef();

  constructor() {
    super();
    this.state = { email: '' };
  }

  validate = () => {
    const { email } = this.state;
    const valid = !/^[a-zA-Z0-9]+(([a-zA-Z0-9]\.)*)+([a-zA-Z0-9]*)+@[a-zA-Z0-9]+(\.+[A-Za-z]+)+$/.test(
      email
    );

    return valid;
  };

  handleEmailChange = evt => {
    this.setState({ email: evt.target.value });
  };

  showSnackbarHandler = () => {
    const { message } = appConfig.joinUsSnackbar;

    this.snackbarRef.current.openSnackBar(message);
  };

  canBeSubmitted = () => {
    const { email } = this.state;
    const isDisabled = this.validate(email);

    return !isDisabled;
  };

  handleSubmit = e => {
    e.preventDefault();
    this.setState({ email: '' });
    const submit = !this.canBeSubmitted();

    return submit;
  };

  render() {
    const { email } = this.state;
    const isDisabled = this.validate(email);
    const error = '* Please enter valid e-mail: sample@samle.com!';

    return (
      <section className="joinus container">
        <div className="joinus__info col-6">
          <h2>news letter</h2>
          <p>join us now to get all news and special offers</p>
        </div>
        <form className="joinus__subscribe col-4" onSubmit={this.handleSubmit}>
          <div className="joinus__fieldset">
            <label htmlFor="email" className="mailfield">
              <i className="far fa-envelope" />
              <input
                id="email"
                type="email"
                placeholder="type your email here"
                value={email}
                required
                onChange={this.handleEmailChange}
              />
            </label>
            <button
              disabled={isDisabled}
              onClick={this.showSnackbarHandler}
              type="submit"
            >
              join us
            </button>
            <Snackbar ref={this.snackbarRef} />
          </div>
          <p className={!isDisabled || !email ? 'msg' : 'msg showerror'}>
            {error}
          </p>
        </form>
      </section>
    );
  }
}
export default JoinUs;
