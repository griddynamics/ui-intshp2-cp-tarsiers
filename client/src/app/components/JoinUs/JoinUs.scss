@import '../../../styles/variables';
@import '../../../styles/mixin';

.joinus {
  @include d-flex();

  width: 100%;
  background-color: $color-white;
  padding: 38px 42px 38px 54px;
  margin-top: 100px;

  &__info {
    position: relative;
    @include d-flex(flex-start) {
      align-items: center;
    }

    & > h2 {
      color: $color-main;
      font-family: $main-font-bold;
      font-size: 18px;
      text-align: center;
      text-transform: uppercase;
    }

    & > p {
      color: $color-dark;
      font-family: $main-font-italic;
      font-size: 14px;
      padding-left: 23px;
      text-align: center;
    }
  }

  &__fieldset {
    @include d-flex();
  }

  &__subscribe {
    .mailfield {
      width: 200px;
      @include d-flex(center) {
        align-items: center;
      }

      & > i {
        color: $color-main;
        padding: 10px;
      }
    }

    input {
      border: none;
      line-height: 2;
      padding: 5px 10px;

      &:focus {
        outline: none;
      }

      &::placeholder {
        color: $color-dark;
        font-family: $main-font-italic;
        font-size: 14px;
        opacity: 0.5;
      }

      &.error {
        border-bottom: 1px dotted $color-main;
      }
    }
    .msg {
      visibility: hidden;
      display: none;
      color: $color-dark;
      font-family: $main-font-italic;
      font-size: 10px;
      padding-top: 10px;

      &.showerror {
        visibility: visible;
        display: block;
      }
    }

    button {
      width: 100px;
      line-height: 50px;
      background-color: $color-main;
      border: none;
      color: $color-white;
      font-family: $main-font-bold;
      font-size: 14px;

      &:disabled {
        background-color: $color-dark;
        opacity: 0.2;
        cursor: not-allowed;
      }

      &:focus {
        outline: none;
      }
    }
  }
}

@media screen and (max-width: $media-large) {
  .joinus {
    width: 90%;

    &__subscribe {
      // width: 40%;
      flex-basis: 40%;
    }
  }
}

@media screen and (max-width: $media-medium) {
  .joinus {
    padding: 38px;
    // padding-right: 25px;
    justify-content: flex-start;

    &__info {
      flex-direction: column;
    }

    &__subscribe {
      flex-basis: 50%;

      button {
        line-height: 40px;
        width: 80px;
      }
    }
  }
}

@media screen and (max-width: $media-tablet) {
  .joinus {
    flex-direction: column;

    &__info {
      margin-bottom: 20px;
    }

    &__subscribe {
      align-self: center;
      flex-basis: 100%;
    }
  }
}

@media screen and (max-width: $media-mobile) {
  .joinus {
    padding: 10px;

    &__fieldset {
      justify-content: center;
    }

    &__subscribe {
      input,
      .mailfield {
        width: 150px;
      }

      .mailfield > i {
        display: none;
      }
    }
  }
}
