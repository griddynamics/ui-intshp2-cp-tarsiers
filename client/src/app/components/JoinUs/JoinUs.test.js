import React from 'react';
import { shallow, mount } from 'enzyme';
import JoinUs from './JoinUs';

describe('<JoinUs />', () => {
  it('should match its snapshot', () => {
    const wrapper = shallow(<JoinUs />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to call handleEmailChange and set state', () => {
    const evt = { target: { value: 'example@mail.com' } };
    const wrapper = shallow(<JoinUs />);

    wrapper.find('#email').simulate('change', evt);
    expect(wrapper.state().email).toBe(evt.target.value);
  });

  it('should be able to call showSnackbarHandler', () => {
    const wrapper = mount(<JoinUs />);
    const instance = wrapper.instance();

    instance.snackbarRef.current.openSnackBar = jest.fn();
    wrapper.update();
    const ref = instance.snackbarRef.current.openSnackBar;

    instance.showSnackbarHandler();
    expect(ref).toBeCalled();
  });

  it('should be able to call handleSubmit', () => {
    const evt = { preventDefault() {} };
    const wrapper = shallow(<JoinUs />);
    const instance = wrapper.instance();

    wrapper.setState({ email: 'not@email' });
    wrapper.update();

    expect(wrapper.state().email).toBe('not@email');
    instance.handleSubmit(evt);
    expect(wrapper.state().email).toBe('');
  });
});
