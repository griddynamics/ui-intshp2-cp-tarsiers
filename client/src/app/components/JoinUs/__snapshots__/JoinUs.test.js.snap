// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JoinUs /> should match its snapshot 1`] = `
<section
  className="joinus container"
>
  <div
    className="joinus__info col-6"
  >
    <h2>
      news letter
    </h2>
    <p>
      join us now to get all news and special offers
    </p>
  </div>
  <form
    className="joinus__subscribe col-4"
    onSubmit={[Function]}
  >
    <div
      className="joinus__fieldset"
    >
      <label
        className="mailfield"
        htmlFor="email"
      >
        <i
          className="far fa-envelope"
        />
        <input
          id="email"
          onChange={[Function]}
          placeholder="type your email here"
          required={true}
          type="email"
          value=""
        />
      </label>
      <button
        disabled={true}
        onClick={[Function]}
        type="submit"
      >
        join us
      </button>
      <Snackbar />
    </div>
    <p
      className="msg"
    >
      * Please enter valid e-mail: sample@samle.com!
    </p>
  </form>
</section>
`;
